pkg/server/snapshots.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"time"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/snapshot"
+	"github.com/golang/glog"
+
+	snapshotstore "github.com/kubernetes-incubator/cri-containerd/pkg/store/snapshot"
+)
+
+// snapshotsSyncer syncs snapshot stats periodically. imagefs info and container stats
+// should both use cached result here.
+// TODO(random-liu): Benchmark with high workload. We may need a statsSyncer instead if
+// benchmark result shows that container cpu/memory stats also need to be cached.
+type snapshotsSyncer struct {
+	store       *snapshotstore.Store
+	snapshotter snapshot.Snapshotter
+	syncPeriod  time.Duration
+}
+
+// newSnapshotsSyncer creates a snapshot syncer.
+func newSnapshotsSyncer(store *snapshotstore.Store, snapshotter snapshot.Snapshotter,
+	period time.Duration) *snapshotsSyncer {
+	return &snapshotsSyncer{
+		store:       store,
+		snapshotter: snapshotter,
+		syncPeriod:  period,
+	}
+}
+
+// start starts the snapshots syncer. No stop function is needed because
+// the syncer doesn't update any persistent states, it's fine to let it
+// exit with the process.
+func (s *snapshotsSyncer) start() {
+	tick := time.NewTicker(s.syncPeriod)
+	go func() {
+		defer tick.Stop()
+		for {
```

Comment:
so how is this go routine going to exit ? should we pass a context and cancel it ? 
Note:  I did see the comment above :)

Reply:
this seems expensive to be running on a timer even if there are no changes.. maybe a TODO to come up with better solution.

Reply:
> so how is this go routine going to exit ? should we pass a context and cancel it ?
Note: I did see the comment above :)

The goroutine will only stop when the process stops. Mentioned in the comment ` No stop function is needed because the syncer doesn't update any persistent states, it's fine to let it exit with the process.`

> this seems expensive to be running on a timer even if there are no changes.. maybe a TODO to come up with better solution.

Yes, this is expensive, but we have to do it. :( This is also how cadvisor works. Kubernetes retrieves stats every <10 second, and expect the stats should be returned immediately. So we have to cache the stats periodically for Kubernetes to use.

Reply:
right.. I remember that.. just pointing out we could use a todo to come up with a better way.

Reply:
@mikebrow OK. Will do.

Reply:
Done.